annotation_pipeline_v2/fdr.py

```diff
@@ -0,0 +1,147 @@
+        merged = merged.sort_values('msm', ascending=False)
+        decoy_cumsum = (merged.is_target == False).cumsum()
+        target_cumsum = merged.is_target.cumsum()
+        base_fdr = np.clip(decoy_cumsum / target_cumsum, 0, 1)
+        base_fdr[np.isnan(base_fdr)] = 1
+        target_fdrs = merged.assign(fdr=base_fdr)[lambda df: df.is_target == 1]
+        target_fdrs = target_fdrs.drop('is_target', axis=1)
+        target_fdrs = target_fdrs.sort_values('msm')
+        target_fdrs = target_fdrs.assign(fdr=np.minimum.accumulate(target_fdrs.fdr))
+        target_fdrs = target_fdrs.sort_index()
+        return target_fdrs
+
+    def merge_rankings(results):
+        mols = pd.concat(results).rename_axis('formula_i').reset_index()
+        return mols.groupby('formula_i').agg({'fdr': np.nanmedian, 'mol': 'first'})
+
+    pw = pywren.ibm_cf_executor(config=config, runtime_memory=2048)
+    data_bucket = input_data['bucket']
+
+    if False:
+        # !! Disabled this branch because PyWren seems to stop receiving results after the first map_reduce task
+        # The map part of the map-reduce is pretty simple - possibly grouping everything into a single task per group would be better
+
+        # Run a separate map-reduce task for each independent set of rankings
+        ranking_futures = []
+        for group_i, group in rankings_df.groupby('group_i'):
+            target_rows = group[group.is_target]
+            decoy_rows = group[~group.is_target]
+
+            for i, target_row in target_rows.iterrows():
+                iterdata = [(f'{data_bucket}/{target_row.key}', data_bucket, decoy_row.key)
+                            for i, decoy_row in decoy_rows.iterrows()]
+                futures = pw.map_reduce(run_ranking, iterdata, merge_rankings)
+                ranking_futures.append((target_row, futures))
+
+        # Get & concat results
+        fdr_dfs = []
+        for target_row, futures in ranking_futures:
+            results = pw.get_result(futures)
+            print(results)
+            result_df = pd.DataFrame({'database_path': target_row.database_path,
+                                      'mol': results.mol,
+                                      'adduct': target_row.adduct,
+                                      'modifier': target_row.modifier,
+                                      'fdr': results.fdr},
+                                     index=results.index)
+            fdr_dfs.append(result_df)
+    else:
+        fdr_dfs = []
+        for group_i, group in rankings_df.groupby('group_i'):
+            target_rows = group[group.is_target]
+            decoy_rows = group[~group.is_target]
+
+            for i, target_row in target_rows.iterrows():
+                iterdata = [(f'{data_bucket}/{target_row.key}', data_bucket, decoy_row.key)
+                            for i, decoy_row in decoy_rows.iterrows()]
+                futures = pw.map_reduce(run_ranking, iterdata, merge_rankings)
```

Comment:
This PyWren map func is executed multiple time, I suggest to execute everything with one PyWren map to reduce times by init rankings_df groups data first (its size is low enough to do so)

Reply:
I've changed this to just a `map`